feat(kad): implement automatic client mode

thomaseizinger opened the pull request:
## Description

<!--
Please write a summary of your changes and why you made them.
This section will appear as the commit message after merging.
Please craft it accordingly.
For a quick primer on good commit mesages, check out this blog post: https://cbea.ms/git-commit/

Please include any relevant issues in here, for example:

Related https://github.com/libp2p/rust-libp2p/issues/ABCD.
Fixes https://github.com/libp2p/rust-libp2p/issues/XYZ.
-->

Currently, the kademlia behaviour can only learn that the remote node supports kademlia on a particular connection if we successfully negotiate a stream to them.

Using the newly introduced abstractions from #3651, we don't have to attempt to establish a stream to the remote to learn whether they support kademlia on a connection but we can directly learn it from the `ConnectionEvent::RemoteProtocolsChange` event. This happens directly once a connection is established which should overall benefit the DHT.

Clients do not advertise the kademlia protocol and thus we will immediately learn that a given connection is not suitable for kadmelia requests. We may receive inbound messages from it but this does not affect the routing table.

Resolves: #2032.

## Notes & open questions

~Draft because it is~ stacked on top of:

- https://github.com/libp2p/rust-libp2p/pull/3651
- https://github.com/libp2p/rust-libp2p/pull/3954
- https://github.com/libp2p/rust-libp2p/pull/3979
- https://github.com/libp2p/rust-libp2p/pull/3980

~Is `Mode::Server` a good default? Should it be `Mode::Client` to be "safer"?
Should we even have an overall mode? With how we are now thinking of supported protocols, we could be a server on one connection and a client on another.~

Depends-On: https://github.com/libp2p/rust-libp2p/pull/3954.

<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Report changes in supported protocols back to `ConnectionHandler`

thomaseizinger committed:
Allow reporting of supported protocols by remote

thomaseizinger committed:
Consume supported protocols in identify

thomaseizinger committed:
Report a remote's protocols to other handlers

thomaseizinger committed:
Implement kademlia client-mode

thomaseizinger committed:
Add test for kademlia client mode

thomaseizinger committed:
Add tests for two servers connecting

thomaseizinger committed:
Report additions and removals of protocols instead

thomaseizinger committed:
Report listen protocols on startup to connection

thomaseizinger committed:
Introduce `SupportedProtocols` type

thomaseizinger committed:
Deduplicate code and propagate supported protocols only once

thomaseizinger committed:
Extend docs

thomaseizinger committed:
Merge branch 'master' into 2680-explore

thomaseizinger committed:
Fix compile errors

thomaseizinger committed:
Report changes to listen/external address set

thomaseizinger committed:
Simplify identify protocol

thomaseizinger committed:
Check for changes in inbound protocols on every poll

thomaseizinger committed:
Fix clippy lints

thomaseizinger committed:
Report changes in `SupportedProtocols`

thomaseizinger committed:
Only report changes to handler if there actually was a change

thomaseizinger committed:
Do not implicitly dial peers upon push

thomaseizinger committed:
Merge branch 'master' into 2680-explore

thomaseizinger committed:
Remove unused import

thomaseizinger committed:
Fix compile error

thomaseizinger committed:
Remove dbg!

thomaseizinger committed:
Update swarm/src/handler.rs

Co-authored-by: Max Inden <mail@max-inden.de>

thomaseizinger committed:
Merge branch '2680-explore' of github.com:libp2p/rust-libp2p into 2680-explore

thomaseizinger committed:
Combine match arms where possible

thomaseizinger committed:
Add comment explaining static hashset

thomaseizinger committed:
Add docs

thomaseizinger committed:
Update supported protocols for push messages

thomaseizinger committed:
Use `pop` to avoid panicking branch in `remove`

thomaseizinger committed:
Use `poll_unpin`

thomaseizinger committed:
Use `if let` for consistency

thomaseizinger committed:
Rewrite to `if let` for consistency

thomaseizinger committed:
Fill in todo in toggle

thomaseizinger committed:
Merge branch 'master' into 2680-explore

thomaseizinger committed:
Merge branch 'master' into 2680-explore

thomaseizinger committed:
Merge branch 'master' into 2680-explore

thomaseizinger committed:
Restore `libp2p-kad`

thomaseizinger committed:
Fix formatting

thomaseizinger committed:
Implement kademlia client-mode

thomaseizinger committed:
Update cargo lock

thomaseizinger committed:
Fix unit tests

thomaseizinger committed:
Change test to assert actual events received

thomaseizinger committed:
Don't report empty set of protocols to handler

thomaseizinger committed:
Use constructor

thomaseizinger committed:
Introduce test helper

thomaseizinger committed:
Make tests less noisy

thomaseizinger committed:
Further simplify test and add comments

thomaseizinger committed:
Add failing test

thomaseizinger committed:
Make test pass, i.e. only report actual changes back to the handler

thomaseizinger committed:
Extract helpers to make fields crate-private and add docs

thomaseizinger committed:
Return `SmallVec` from `from_full_sets` which allows iteration

thomaseizinger committed:
Fix clippy warnings

thomaseizinger committed:
Fix rustdoc

thomaseizinger committed:
Merge branch '2680-explore' into feat/kademlia-automatic-client-mode

thomaseizinger committed:
Only update protocol support on change message

thomaseizinger committed:
Update docs

thomaseizinger committed:
Merge branch 'master' into feat/kademlia-automatic-client-mode

thomaseizinger committed:
Minimize diff

thomaseizinger committed:
More resilient pattern match

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -399,6 +402,14 @@ impl KademliaConfig {
+    pub fn set_mode(&mut self, m: Mode) -> &mut Self {
+        self.mode = m;
+        self
+    }
```
The `Mode` needs to be changeable at runtime. I expect the most common usage to be a call to `set_mode` based on an at-runtime `libp2p-autonat` result.

mxinden commented on protocols/kad/src/handler_priv.rs:
```diff
@@ -544,7 +547,7 @@ where
         };
 
-        if let ProtocolStatus::Unconfirmed = self.protocol_status {
+        if let ProtocolStatus::Unknown = self.protocol_status {
```
Currently the `ProtocolStatus` is indifferent on whether the protocol support of the remote is outbound (client) or inbound (server) or both. I don't think there is any value in knowing whether the remote node is a client or not from the perspective of the local `ConnectionHandler`. The only thing that matters to `libp2p-kad` is whether the remote node is a server or not. Thus I don't think there is any value in switching from `ProtocolStatus::Unknown` to `ProtocolStatus::Cofirmed` in `on_fully_negotiated_inbound`. Does that make sense?

thomaseizinger committed:
Merge branch 'master' into feat/kademlia-automatic-client-mode

thomaseizinger committed:
Initial refactoring to new events

thomaseizinger committed:
Generate candidates based on address translation

thomaseizinger committed:
Use candidates in AutoNAT

thomaseizinger committed:
Only keep minimal external address state in swarm

thomaseizinger committed:
Merge branch 'master' into feat/kademlia-automatic-client-mode

thomaseizinger committed:
Always enable server mode for inbound connections

thomaseizinger committed:
Determine mode based in inbound/outbound connection

thomaseizinger committed:
Update identify docs

thomaseizinger committed:
Insert and remove external addresses

thomaseizinger committed:
Update docs

thomaseizinger committed:
Apply suggestions from code review

Co-authored-by: Max Inden <mail@max-inden.de>

thomaseizinger committed:
Fix doc link

thomaseizinger committed:
Remove private NAT status test

We no longer test external addresses.

thomaseizinger committed:
Fix more doc links

thomaseizinger committed:
Merge branch 'master' into 2032/external-addr

thomaseizinger committed:
Fix compile error with latest master

thomaseizinger committed:
Delete tests

thomaseizinger committed:
Update docs

thomaseizinger committed:
Revert "Delete tests"

This reverts commit 62ec4828ca9ea7a293b57cec0b11d697f39eed72.

thomaseizinger committed:
Fix tests by adding a new API to AutoNAT

thomaseizinger committed:
Rename to `probe_address`

thomaseizinger committed:
Merge branch '2032/external-addr' into feat/kademlia-automatic-client-mode

thomaseizinger committed:
Merge branch 'master' into 2032/external-addr

thomaseizinger committed:
Merge branch '2032/external-addr' into feat/kademlia-automatic-client-mode

thomaseizinger committed:
Split handler ctor into inbound and outbound

thomaseizinger committed:
Remove handler config

thomaseizinger committed:
Outbound connections are never in server mode

thomaseizinger committed:
Add log for outbound client mode

thomaseizinger committed:
Activate server mode for inbound connections based on our external addresses

thomaseizinger committed:
Push updated protocols to peer

thomaseizinger committed:
Fix starved connection handler on local-protocols change event

thomaseizinger committed:
Add test for automatic switch to server mode on existing connections

thomaseizinger committed:
Reconfigure existing connections to server mode

thomaseizinger committed:
Add more logs for routing table updates

thomaseizinger committed:
Include reported in condition

thomaseizinger committed:
Fix formatting

thomaseizinger committed:
Remove log that we can't guarantee

thomaseizinger committed:
Only log if we are actually connected

thomaseizinger commented on protocols/kad/src/handler.rs:
```diff
@@ -468,16 +476,88 @@ pub struct KademliaRequestId {
+                log::debug!("Operating in server-mode on new inbound connection from peer {remote_peer_id} under the assumption that one of [{confirmed_external_addresses}] routes to {local_addr}");
+
+                Mode::Server
+            }
```
This is where we are making assumptions. A few things we could do is:

- Compare the protocols of the local and remote address. If our confirmed external address is TCP and the remote dialed us via QUIC, then we shouldn't advertise kademlia on this connection.
- More generally, it would be good to track a mapping of external addresses and which local interface they belong to, then we could accurately decide here.

mxinden replied:
In my eyes, for the sake of simplicity, our first iteration should have the same `Mode` across all connections (as implemented above). I don't feel comfortable with the suggested matching of local and external address, especially when it comes to compatibility across versions and implementations.

As a downside, this might lead to the local node being added to remote routing tables even though the remote node can not connect to the local node. The remote node will eventually notice and remove the local node from its routing table. Kademlia's routing table mechanism of keeping the oldest online nodes helps us here.

thomaseizinger replied:
This version has the same `Mode` across all connections. I send a `ReconfigureMode` message to all inbound connections. This matching here only differs in the log message. We don't inspect the addresses themselves at all. The logic essentially is:

- Do we have at least 1 external address? Switch into server mode.
- Do we have no external addresses? Switch into client mode.

The important bit about this log message is that we are telling the user that we are making an assumption about their network configuration here.

mxinden replied:
I think we are on the same page here, there just seems to be a miscommunication.

> A few things we could do is:

I am suggesting to not do the initial suggestions, but to keep the logic as is. The only change I would like to see is that we treat outbound connections just like inbound connections. More specifically, when changing to `Mode::Server`, that we do so on outbound connections as well. See https://github.com/libp2p/rust-libp2p/pull/3877/files#r1204946875.

thomaseizinger committed:
Merge branch 'master' into feat/kademlia-automatic-client-mode

thomaseizinger committed:
Set listen address as external address to trigger server mode

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2419,7 +2424,28 @@ where
+                            external_addresses: self.external_addresses.iter().cloned().collect(),
+                        },
+                    }
+                }));
```
Why do we only reconfigure the mode of inbound connections?

The remote node of an outbound connection might be aware of our external address through some other means, e.g. `libp2p-identify`. Thus, us switching to server mode, is of interest to the remote, as they can now add us to their routing table.

thomaseizinger replied:
How would the remote know whether we are reusing ports and are thus reachable on this observed address?

Are you arguing that the policy of not adding outbound connections to the routing table is internal to `rust-libp2p` and should not be assumed for other nodes?

mxinden replied:
> Are you arguing that the policy of not adding outbound connections to the routing table is internal to `rust-libp2p` and should not be assumed for other nodes?

Yes, that is my understanding.

mxinden commented on protocols/kad/src/handler.rs:
```diff
@@ -468,16 +476,88 @@ pub struct KademliaRequestId {
+        peer: PeerId,
+    ) -> Self {
+        // Outbound connections are always in client-mode because we may be using an ephemeral port for dialing and thus the observed address may not be reachable.
+        log::debug!("Operating in client-mode on new outbound connection to peer {peer} at {addr}");
```
I am not sure this is a good idea. Does the go-libp2p Kademlia implementation see the `Mode` as a per-connection property or as a property across all connections? For the sake of consistency and reducing complexity, I suggest treating the `Mode` as a property across all connections. In other words, either the node with all its connections is in `Mode::Server` or `Mode::Client`. Whether the remote node of an outbound connection adds us to their routing table is up to them.

One example would be, what if the local node advertises `Mode::Client` on this connection, but advertises the Kademlia protocol via `libp2p-identify`. How do other implementations cope with this seemingly inconsistent information?

We can still role out this optimization in the future. Though for now, I find treating the `Mode` as a global property much easier to reason about.

thomaseizinger replied:
Let's discuss this in [this](https://github.com/libp2p/rust-libp2p/pull/3877#discussion_r1204946875) thread.

mxinden commented on protocols/kad/src/handler.rs:
```diff
@@ -680,6 +759,45 @@ impl ConnectionHandler for KademliaHandler {
+                            .iter()
+                            .map(|addr| addr.to_string())
+                            .collect::<Vec<_>>()
+                            .join(", ");
```
We should only do the `collect` in case the log level is `debug`, right?

mxinden commented on protocols/kad/src/handler.rs:
```diff
@@ -760,8 +878,38 @@ impl ConnectionHandler for KademliaHandler {
+                        .iter()
+                        .any(|p| self.protocol_config.protocol_names().contains(p));
+
+                    match (remote_supports_our_kademlia_protocols, self.protocol_status) {
+                        (true, ProtocolStatus::Confirmed | ProtocolStatus::Reported) => {}
+                        (true, _) => {
+                            log::info!(
+                                "Remote {} now supports our kademlia protocol",
+                                self.remote_peer_id
+                            );
+
+                            self.protocol_status = ProtocolStatus::Confirmed;
+                        }
+                        (false, ProtocolStatus::Confirmed | ProtocolStatus::Reported) => {
+                            log::info!(
+                                "Remote {} no longer supports our kademlia protocol",
+                                self.remote_peer_id
+                            );
+
+                            self.protocol_status = ProtocolStatus::NotSupported;
+                        }
+                        (false, _) => {}
+                    }
```
:rocket: this is very clean!

thomaseizinger committed:
Merge branch 'master' into feat/kademlia-automatic-client-mode

mxinden reviewed (commented):
In my eyes we should treat in- and outbound connections the same. We can still differentiate as an optimization in future iterations.

No other objections to this pull request.

thomaseizinger committed:
Merge branch 'master' into feat/kademlia-automatic-client-mode

thomaseizinger committed:
Have a global mode for kademlia

thomaseizinger committed:
Only allocate in debug mode

thomaseizinger committed:
Inline `limit` field

thomaseizinger committed:
Rewrite `ExternalAddresses` to be backed by `Vec`

This allows us to expose an `as_slice` function.

thomaseizinger committed:
Move mode variable to `Behaviour`

thomaseizinger committed:
Ensure newer external addresses are prioritized

thomaseizinger committed:
Add failing test for prioritizing addresses

thomaseizinger committed:
Make use of newer addresses and optimize insertion

thomaseizinger committed:
Remove dead code

mxinden commented on protocols/kad/CHANGELOG.md:
```diff
@@ -6,7 +6,13 @@
 
+- Automatically configure client/server mode based on external addresses.
+  If we have or learn about an external address of our node, we operate in server-mode and thus allow inbound requests.
+  By default, all connections are in client-mode and only allow outbound requests.
```
```suggestion
  By default, a node is in client-mode and only allows outbound requests.
```

Since the `Mode` is a global setting now.

thomaseizinger committed:
Always change mode

mxinden commented on protocols/kad/src/handler.rs:
```diff
@@ -559,7 +576,7 @@ impl KademliaHandler {
         }
 
-        debug_assert!(self.config.allow_listening);
+        debug_assert!(self.mode == Mode::Server);
```
The `ConnectionHandler` might have switched to `Mode::Client` in the meantime. That would not be an invalid state transition, right?

thomaseizinger replied:
Uhm interesting, I guess that is true. I'll remove it.

thomaseizinger committed:
Move debug assert

thomaseizinger committed:
Update protocols/kad/CHANGELOG.md

Co-authored-by: Max Inden <mail@max-inden.de>

thomaseizinger committed:
Remove debug assert

We could have switched to client mode between requesting the stream and establishing
it.

mxinden commented on swarm/src/behaviour/external_addresses.rs:
```diff
@@ -29,18 +18,40 @@ impl ExternalAddresses {
-                if self.addresses.len() < self.limit {
-                    return self.addresses.insert((*addr).clone());
+                if self.addresses.contains(addr) {
+                    return false;
```
In case the address is already here, shouldn't we move it to the back such that fresh addresses are less likely to be removed? Are `NetworBehaviour`s supposed to continuously report the address?

thomaseizinger replied:
Done.

thomaseizinger committed:
Fix formatting

mxinden commented on swarm/src/behaviour/external_addresses.rs:
```diff
@@ -29,18 +18,40 @@ impl ExternalAddresses {
+                }
+
+                if self.addresses.len() >= MAX_LOCAL_EXTERNAL_ADDRS {
+                    // If we hit the limit, we can add the address in O(1):
```
Just so that we are on the same page. The `self.addresses` `Vec` doesn't grow beyond `<100` addresses. Also I would expect change frequency to this datastructure on the order of `> 1s`. Thus I don't think even an `O(n^2)` would be a problem here. Would you disagree?

thomaseizinger replied:
Removed.

mxinden commented on swarm/src/behaviour/external_addresses.rs:
```diff
@@ -29,18 +18,40 @@ impl ExternalAddresses {
+                if self.addresses.len() >= MAX_LOCAL_EXTERNAL_ADDRS {
+                    // If we hit the limit, we can add the address in O(1):
+                    self.addresses.push((*addr).clone());
+                    self.addresses.swap_remove(0);
```
Can you add a `log::debug` line here? I would expect this to be hard to debug otherwise.

thomaseizinger replied:
Ah, this isn't as clean as I thought it is because we are not removing the oldest entry ...

thomaseizinger committed:
Keep the list of addresses in insertion order

thomaseizinger committed:
Refresh newly reported addresses

mxinden commented on swarm/src/behaviour/external_addresses.rs:
```diff
@@ -29,24 +18,60 @@ impl ExternalAddresses {
+
+                if self.addresses.len() > MAX_LOCAL_EXTERNAL_ADDRS {
+                    self.addresses.pop();
                 }
```
I still think this needs a log line. Otherwise I would be surprised if someone can easily debug why their previously confirmed external address is no longer part of the set. Any objections?

thomaseizinger replied:
Previously, we didn't have a log line either so I wouldn't consider this a regression.

Happy to add it in a follow-up!

thomaseizinger replied:
I guess you are saying that previously we just stopped adding them and now we remove them from the set, I see.

Will add a log line.

mxinden reviewed (approved):
Ready to go apart from the smaller suggestion below.

thomaseizinger committed:
Merge branch 'master' into feat/kademlia-automatic-client-mode

thomaseizinger committed:
Add log for dropping previously confirmed external address

folex commented on protocols/kad/src/handler.rs:
```diff
@@ -368,6 +365,9 @@ pub enum KademliaHandlerIn {
     /// for the query on the remote.
     Reset(KademliaRequestId),
 
+    /// Change the connection to the specified mode.
```
What's the effect of that? Is there a network message sent all my connections when I learn about external_addresses? 

I think it makes sense to elaborate on that in the comment. Describe effects of other peers receiving this message.

Also, isn't that a lot of overhead? What if my addresses go on and off often? Not sure why that might happen, but now how do I make sure that doesn't happen?

thomaseizinger replied:
Nobody receives any messages. Client vs server mode essentially just means, "do we accept inbound kademlia streams" which translates to `rust-libp2p` as "do we offer kademlia via `ConnectionHandler::listen_protocols`".

folex commented on protocols/kad/src/handler.rs:
```diff
@@ -601,11 +617,9 @@ impl ConnectionHandler for KademliaHandler {
-                .map_upgrade(Either::Left)
-        } else {
-            SubstreamProtocol::new(Either::Right(upgrade::DeniedUpgrade), ())
+        match self.mode {
```
Can you please add some comments on what happens here?

That sounds like a rather impactful thing. Now Peer can allow and disallow accepting connections _on its own_, and I as a developer have no control over that.

folex replied:
_I'll try to outline my understanding of the effects of this change. I'm certainly can be wrong here, so forgive me if I am :)_

So, previously if I had `allow_listening = true`, my peer would accept incoming connections even before `Identify` happened.

Now Kademlia handshake behaviour depends on whether `Identify` already happened or not. So instead of 1 handshake on Kademlia protocol and 1 handshake on Identify, we now have (depending on handshakes order) several scenarios:

PeerA connects to PeerB. PeerA is in Server mode, PeerB has not yet learned her external addresses, so it is in Client mode.

### Scenario 1: failed connection
1. PeerA tries to connect to PeerB
2. PeerA attempts Kademlia handhsake with PeerB, PeerB rejects because Mode::Client => upgrade::DeniedUpgrade
3. PeerA performs a succesful handshake via Identify with PeerB
4. PeerB learns its external address, switches to Server mode
5. PeerB sends `ReconfigureMode` to _all its kademlia connections_, but PeerA is not among them
6. Kademlia handhsake between A and B has failed
7. Connection failed

If PeerA tries again, connection will succeed, even though _nothing has changed_ in the setup. Technically, nothing prevents PeerB of accepting PeerA's connection the first time. But now PeerA has to retry if she's sure she brought external_addresses to PeerB.

### Scenario 2: succesful connection
1. PeerA tries to connect to PeerB 
2. PeerA performs a succesful handshake via Identify with PeerB
3. PeerB switches to Server mode
4. PeerA attempts Kademlia handshake with PeerB
5. PeerB accepts
6. Connection succesful

If I understand correctly, there's no strict order in which handshakes happen, right? So it is basically a race condition.

thomaseizinger replied:
> PeerA attempts Kademlia handhsake with PeerB, PeerB rejects because Mode::Client => upgrade::DeniedUpgrade

Can you elaborate on how that handshake is triggered? At least `rust-libp2p` does not attempt to speak the kademlia protocol to a peer that did not advertise it first. This is the whole point of this PR: By filtering peers based on whether the advertise kademlia, we get healthier routing tables and as a result, kademlia queries finish a lot quicker. See https://github.com/libp2p/rust-libp2p/pull/3877#issuecomment-1574940336.
